### ISSUE TITLE
Add VRx Protocol for basic (multi-)receiver device control

### ISSUE THREAD

**Issue by sl33pydog** [RotorHazard/RotorHazard#369]
This would be an enchancment for race organizers that have extra analogue video rxs and want to bring them into a broadcast/recording software but want them to update automatically hands off from the main node when you change it on the settings page.

After discussing it on the Facebook page something similar to the Wizardrx from Propnuts but using an esp32 could be used for this.  What the esp32 would do is change the frequencies of each rx5808 or equivalent using 10 gpio pins.  Mosi and slck are tied to each receiver and each receiver would have its own SS pin.  See picture of an example.  I propose to use the Wroom Esp32 Dev board since it pretty much has everything on it and is fairly inexpensive.

As a side note this could expand into a general purpose gate effect node as you could have an if rssi is triggered then trigger something situation since the Wroom has more than 10 ADC connection.  But this is secondary and probably already in the chorus32 Integration.
![IMG_20200712_154601](https://user-images.githubusercontent.com/61440555/87323428-5c6e0580-c4f4-11ea-88ab-5ac6dfe758ca.jpg)

**Comment by HazardCreative** [comment]
This sounds like a perfect application for the [VRx Control functionality](https://github.com/RotorHazard/RotorHazard/blob/master/doc/Video%20Receiver.md) that's just been added in 2.2. We can add a dedicated protocol for sending commands to this device via MQTT. Get your ESP32 running with a control connection and MQTT, (The ESP section of [clearview_interface_public](https://github.com/ryaniftron/clearview_interface_public/tree/master/src/cvesp32) is a great starting point,) and we can use this issue to discuss the protocol specifics as you get further along.

**Comment by sl33pydog** [comment]
@HazardCreative I've been able to build it on an ESP32 Wroom and get to the config page.  I'm assuming this gives it a control connection and MQTT.

**Comment by HazardCreative** [comment]
What have you built on the ESP32 so far? Do you have control of your receivers? Any UI for that? Does it connect to MQTT?

@ryaniftron next steps on our end will be the protocol expansion you've been working on.

**Comment by sl33pydog** [comment]
@HazardCreative I haven't adding anything to it so far.  I need to set up the mqtt on the pi.  From looking around the iftron wiki it looks like I would need an Arduino to talk to the esp32 with how it's set up now.  If that's true then the easiest way would be for the code from the wizardrx could be forked and add a communication protocol to it and then variables that could be pushed from the esp32.  I would assume the more difficult way would be to write new code that turns the esp32 into the channel changer like the circuit I showed above.  Either way is doable for me on the hardware side (more so since I have a wizardrx lying around) but I am very limited when it comes to writing out code.  If I know what I'm supposed to add to the wizard Rx code I could test that and see if I can get 1 of the nodes to change channels based on the esp32 talking over tx/Rx to the Arduino nano on the wizardrx.  Thoughts?

**Comment by HazardCreative** [comment]
Why would an Arduino be needed? I guess I don't understand what you have built so far. I assumed when you said you had built it, you meant as the diagram above with the ESP directly controlling the RX modules.

**Comment by sl33pydog** [comment]
@HazardCreative sorry I should have clarified.  I have a the esp32 wroom on a breadboard and some rx5808s ready to add to it like how it is on the diagram.  When I looked at how the clearview interface was written I saw that it sent commands to their receiver on the tx Rx lines.  I wasn't sure if you guys wanted to keep it that way and then we could just add an Arduino that would do the channel control but if it can be built all on the Esp32 Wroom then that would be ideal.  I was not sure if you wanted a universal node that would talk to other systems or have it all available using the Esp32

**Comment by ryaniftron** [comment]
> What have you built on the ESP32 so far? Do you have control of your receivers? Any UI for that? Does it connect to MQTT?
> 
> @ryaniftron next steps on our end will be the protocol expansion you've been working on.

We need to define the message responses and parse them here. 
https://github.com/ryaniftron/RotorHazard/blob/dev_rcvr/doc/MQTT%20API.md

I can go ahead and flesh out the ones we need for normal VRX control and responses once the CVCM code is written to actually form those responses. 

@sl33pydog Feel free to post any custom messages or things you need defined for the MQTT JSON messages and I can add it to the VRxController. 

In reference to your hardware, I would recommend referencing clearview_interface_public as another subrepository and writing your SPI control that way you can continue to receive updates to the WIFI and MQTT drivers that we are working on for the ESP32. Yes, we do plan to keep the CVCM code using UART because the CV2.0 only has UART capability. Yes, I'd recommend building all of it into the ESP32 - better to write WIFI JSON -> SPI on ESP32 rather than WIFI JSON -> Serial ESP -> Serial Arduino -> SPI on Arduino. Adding another arduino will only be more complex from both a hardware and software perspective. 

Let me know if you need some additional help with this. Michael and I have already figured out how to use the CVCM code in a different repo to talk to a different device. It's in progress, but we can share with you the necessary architecture.

**Comment by sl33pydog** [comment]
So bear with me on this.  I have no code experience.  I typically just take code and piece it together and through trial and error try to get it to work.  So I may be asking a ton of questions.  Here is where I've placed the code that can flash to the esp32 wroom through arduino ide and change 2 different receivers but only statically (aka reupload code with different frequencies).

https://github.com/sl33pydog/wizardrx/tree/sl33pydog-patch-1/Source/

I thought a good starting point would be to fork the WizardRX code to esp32, however it assumes you are sending it the 4 number frequency (like 5800) to it.  I thought it would be fairly simple compared to looking at Chorus32 for spi control, even though it's meant for an esp32 but again they use arduino to flash.

So I think I'm at a good stopping point to ask what should I do now.  How do I translate C++ to C and inject it into your code @ryaniftron.  I figured that what I could do is take your drop down menus for the test page for "Channel" and change them to frequencies that I can feed into the code I've adjusted to see if it will change channels on the fly.  Then we could replace that with either a channel table or maybe have the MQTT be 4 digit frequencies.  Once I can figure out how to change the code from C++ to C to use, I'll then fork the clearview_interface_public so I can push and pull.

Let me know if I'm helping or not.  I'm learning as I'm going so any critiquing would also help.

**Comment by sl33pydog** [comment]
Got it to work.  It's taken me a while to learn how to make it but most of this was cobbled together from tutorials and the wizardRX platform.  It's not pretty or efficient but it works.  If there is a user name and pw it will need to be added into the code.  It's not written for one.  I know @ryaniftron you mentioned that we're moving to json.  If so I'll need to add a parser and would need to know the chain.

For now it works with MQTT messages.  Just need to change the topics in the beginning.  The only bad part is that if others are on the network they could change the frequencies by publishing on the same topics.

https://github.com/sl33pydog/MQTT-Multi-Video-Receiver-SPI-Controller

**Comment by HazardCreative** [comment]
Public access to the MQTT server is an issue, but not one you should need to solve. It should really only need secured on the server-side as there are no write commands back to the server. RotorHazard needs built out to support MQTT auth, connected devices should just be able to listen.

**Comment by sl33pydog** [comment]
> 
> 
> Public access to the MQTT server is an issue, but not one you should need to solve. It should really only need secured on the server-side as there are no write commands back to the server. RotorHazard needs built out to support MQTT auth, connected devices should just be able to listen.

Something like an ACL where the topics are read only and the only one who can publish is localhost?  Something like example 3 on this site? http://www.steves-internet-guide.com/topic-restriction-mosquitto-configuration/

**Comment by ryaniftron** [comment]
> Got it to work. It's taken me a while to learn how to make it but most of this was cobbled together from tutorials and the wizardRX platform. It's not pretty or efficient but it works. If there is a user name and pw it will need to be added into the code. It's not written for one. I know @ryaniftron you mentioned that we're moving to json. If so I'll need to add a parser and would need to know the chain.
> 
> For now it works with MQTT messages. Just need to change the topics in the beginning. The only bad part is that if others are on the network they could change the frequencies by publishing on the same topics.
> 
> https://github.com/sl33pydog/MQTT-Multi-Video-Receiver-SPI-Controller

Yes, I'm still wrapping up that PR. The publishers on RH are done, the publishers and subscribers on ESP32 are done, and last up is the subscribers on RH (my goal this week). The API I have so far is published here:
https://github.com/ryaniftron/clearview_interface_public/blob/master/src/cvesp32/docs/cv_api.md

That should be plenty for you to work on for the time being. 

Yes, we will need access rules for sure in the future once devices have different permissions.

**Comment by HazardCreative** [comment]
Note that once the CVCM is done, we can set up a modified control set for your purposes. For example, the CV is only one ESP (WiFi) device and one RX, but you have one WiFi controller and multiple RX. We may be able to make tweaks to help simplify and clean that up for you. Getting the JSON parser running will be a necessary step regardless. There are libraries for the bulk of that work such as ArduinoJson.

**Comment by ryaniftron** [comment]
Yep. Feel free to check out the CVCM code as we already are using the cJSON library included with ESPIDF. You can even add the clearview_interface_public/src/cvesp32/components as git submodule for your project and leverage the work we have already done for an ESP32. 
@HazardCreative Did this for rapidfire control to use the WiFi and web page infrastructure I wrote.

**Comment by HazardCreative** [comment]
System now supports plugins for the VRx Control subsystem. #721
